File: android/app/src/main/java/com/rbdarts/feature/standalonegame/StandaloneGameViewModel.kt
package com.rbdarts.feature.standalonegame

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.rbdarts.core.data.InMemoryRBDartsRepository
import com.rbdarts.core.domain.GameSummary
import com.rbdarts.core.domain.ScoringRules
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class StandaloneGameUiState(
    val playerNames: List<String> = listOf("Player 1", "Player 2"),
    val gameName: String = "",
    val currentInning: Int = 1,
    val session: ActiveScoreSession? = null,
    val summary: GameSummary? = null,
    val errorMessage: String? = null
) {
    val liveSummary: GameSummary?
        get() = session?.let {
            ScoringRules.summarizeStandalone(
                participants = it.participants,
                regulationInnings = it.game.regulationInnings,
                extraInningsEnabled = it.game.extraInningsEnabled
            )
        }
}

class StandaloneGameViewModel(
    private val service: StandaloneGameService = StandaloneGameService(InMemoryRBDartsRepository()),
    private val diagnostics: StandaloneGameDiagnostics = StandaloneGameDiagnostics()
) : ViewModel(), StandaloneScoreEntryActions {
    private val _state = MutableStateFlow(StandaloneGameUiState())
    val state: StateFlow<StandaloneGameUiState> = _state

    init {
        viewModelScope.launch {
            service.recoverMostRecentSession()?.let { session ->
                diagnostics.recordRecovery()
                _state.update { it.copy(session = session) }
            }
        }
    }

    fun updateGameName(gameName: String) {
        _state.update { it.copy(gameName = gameName) }
    }

    fun updatePlayerName(index: Int, name: String) {
        _state.update { state ->
            state.copy(playerNames = state.playerNames.toMutableList().also { it[index] = name })
        }
    }

    fun addPlayer() {
        _state.update { it.copy(playerNames = it.playerNames + "Player ${it.playerNames.size + 1}") }
    }

    fun startGame() {
        viewModelScope.launch {
            runCatching {
                service.createGame(
                    name = state.value.gameName,
                    createdByUserId = "local-user",
                    participantNames = state.value.playerNames
                )
            }.onSuccess { session ->
                _state.update { it.copy(session = session, currentInning = 1, summary = null, errorMessage = null) }
            }.onFailure { error ->
                _state.update { it.copy(errorMessage = error.message) }
            }
        }
    }

    override fun enterScore(score: Int, participantId: String) {
        val session = state.value.session ?: return
        viewModelScope.launch {
            runCatching {
                service.enterScore(score, participantId, state.value.currentInning, session)
            }.onSuccess { updated ->
                diagnostics.recordScoreAccepted(state.value.currentInning)
                _state.update { it.copy(session = updated, errorMessage = null) }
            }.onFailure { error ->
                diagnostics.recordInvalidScore(state.value.currentInning)
                _state.update { it.copy(errorMessage = error.message) }
            }
        }
    }

    override fun undo(participantId: String) {
        val session = state.value.session ?: return
        viewModelScope.launch {
            _state.update {
                it.copy(session = service.undoScore(participantId, state.value.currentInning, session))
            }
        }
    }

    override fun advanceInning() {
        _state.update { it.copy(currentInning = it.currentInning + 1) }
    }

    override fun completeGame() {
        val session = state.value.session ?: return
        viewModelScope.launch {
            runCatching { service.complete(session) }
                .onSuccess { summary ->
                    diagnostics.recordCompletedGame()
                    _state.update { it.copy(summary = summary, errorMessage = null) }
                }
                .onFailure { error -> _state.update { it.copy(errorMessage = error.message) } }
        }
    }
}

File: ios/RBDarts/Features/StandaloneGame/StandaloneGameViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
public final class StandaloneGameViewModel {
    public var playerNames: [String] = ["Player 1", "Player 2"]
    public var gameName: String = ""
    public var currentInning: Int = 1
    public var session: ActiveScoreSession?
    public var summary: GameSummary?
    public var errorMessage: String?
    public var isSaving: Bool = false

    private let service: StandaloneGameService
    private let diagnostics: StandaloneGameDiagnostics

    public init(
        service: StandaloneGameService = StandaloneGameService(),
        diagnostics: StandaloneGameDiagnostics = StandaloneGameDiagnostics()
    ) {
        self.service = service
        self.diagnostics = diagnostics
        recover()
    }

    public func startGame() {
        Task {
            do {
                isSaving = true
                session = try await service.createGame(
                    name: gameName,
                    createdByUserId: "local-user",
                    participantNames: playerNames
                )
                currentInning = 1
                summary = nil
                errorMessage = nil
            } catch {
                errorMessage = error.localizedDescription
            }
            isSaving = false
        }
    }

    public func enterScore(_ score: Int, for participantId: String) {
        guard let session else { return }
        Task {
            do {
                self.session = try await service.enterScore(
                    score,
                    participantId: participantId,
                    inningNumber: currentInning,
                    in: session
                )
                diagnostics.recordScoreAccepted(inning: currentInning)
                errorMessage = nil
            } catch {
                diagnostics.recordInvalidScore(inning: currentInning)
                errorMessage = error.localizedDescription
            }
        }
    }

    public func undo(participantId: String) {
        guard let session else { return }
        Task {
            do {
                self.session = try await service.undoScore(
                    participantId: participantId,
                    inningNumber: currentInning,
                    in: session
                )
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }

    public func advanceInning() {
        currentInning += 1
    }

    public func completeGame() {
        guard let session else { return }
        Task {
            do {
                summary = try await service.complete(session)
                diagnostics.recordCompletedGame()
                errorMessage = nil
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }

    public var liveSummary: GameSummary? {
        guard let session else { return nil }
        return ScoringRules.summarizeStandalone(
            participants: session.participants,
            regulationInnings: session.game.regulationInnings,
            extraInningsEnabled: session.game.extraInningsEnabled
        )
    }

    private func recover() {
        Task {
            do {
                session = try await service.recoverMostRecentSession()
                if session != nil {
                    diagnostics.recordRecovery()
                }
            } catch {
                errorMessage = error.localizedDescription
            }
        }
    }
}
